src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/TagHelperDeltaResult.cs

```diff
@@ -10,16 +10,21 @@ namespace Microsoft.CodeAnalysis.Remote.Razor
     internal record TagHelperDeltaResult(
         bool Delta,
         int ResultId,
-        IReadOnlyList<TagHelperDescriptor> Added,
-        IReadOnlyList<TagHelperDescriptor> Removed)
+        IReadOnlyCollection<TagHelperDescriptor> Added,
+        IReadOnlyCollection<TagHelperDescriptor> Removed)
     {
-        public IReadOnlyList<TagHelperDescriptor> Apply(IReadOnlyList<TagHelperDescriptor> baseTagHelpers)
+        public IReadOnlyCollection<TagHelperDescriptor> Apply(IReadOnlyCollection<TagHelperDescriptor> baseTagHelpers)
         {
             if (Added.Count == 0 && Removed.Count == 0)
             {
                 return baseTagHelpers;
             }
 
+            // We're specifically choosing to create a List here instead of an alternate type like HashSet because
+            // results that are produced from `Apply` are typically fed back into two different systems:
+            //
+            // 1. This TagHelperDeltaResult.Apply where we don't iterate / Contains check the "base" collection.
```

Comment:
This stipulation seems likely so change, maybe note in that code that other areas are relying on it to not iterate/Contains?

Reply:
It's a nice idea but I'm not sure I want the icky detail of this to bleed into other components 😢